Double linked list

Robin-Bjornsdotter opened the pull request.

Commit:
Added .so file

Commit:
Added functions and tests:
list_create - allocates memory for the head
list_free - frees allocated memory
list_print
list_prepend - adds an element to the beginning
list_append - adds an element to the end
list_pop - removes lat element
list_reverse - reverses the list
list_get_el - returns an element on the chosen position

test_list_prepend
test_list_append
test_list_pop
test_list_reverse
test_list_get_el

Commit:
Added makefile

disconnect3d committed:
Fixed makefile

There was a copy-paste like bug in the makefile so that `test.o`
directive compiled `tmain.c` instead of `test.c` file ... ;)

Commit:
Small fix of makefile

Commit:
Added head and tail in list structure
First attempt to adjust everything to this

Commit:
Changed double_linked_list.c
Attempt to make it fit to implementation with head and tail

Commit:
Fixed list_append, list_prepend and tests for them

disconnect3d commented on double_linked_list/double_linked_list.c:
```diff
@@ -0,0 +1,129 @@
+
+list* list_create(){
+    list* l = (list*)malloc(sizeof(list));
+    l->quantity = 0;
```
Both `head` and `tail` pointers are uninitialized. Because of that, most of the functions many functions may fail to work properly because they are based on those uninitialized values...

disconnect3d commented on double_linked_list/double_linked_list.c:
```diff
@@ -0,0 +1,129 @@
+            first = first->next;
+            free(second);
+    }
+    l = NULL;
```
This is redundant (it changes just the function local variable to `NULL`, why would we need that?). Instead `free` the list instance ;).

Commit:
Added initialization of head and tail pointers in list_create
Changed list_free according to the comment
list_pop and list_get_el seem to be fine

Commit:
Fixed list_reverse

Commit:
Changed makefile as before it did not create main

disconnect3d commented on double_linked_list/double_linked_list.c:
```diff
@@ -0,0 +1,133 @@
+#include "double_linked_list.h"
+
+list* list_create(){ //how to tell if malloc worked?
+    list* l = (list*)malloc(sizeof(list));
```
Return `NULL` if malloc fails.

disconnect3d commented on double_linked_list/double_linked_list.c:
```diff
@@ -0,0 +1,133 @@
+
+void list_free(list* l){
+    node* first = l->head;
+    while (first->next) {
```
If list is empty (`l->head == NULL`) this code booms.

disconnect3d commented on double_linked_list/double_linked_list.c:
```diff
@@ -0,0 +1,133 @@
+    int ret_el = tmp->data;
+    free(tmp);
+    if (l->quantity < 2)
+        l->tail = l->head;
```
If there is only 1 element in the list, we will get double free (if we call `list_free` afterwards).

Commit:
Fixed list_pop and list_free
Added testcase for popping an element from one element list

Commit:
Fixed `list_pop` so that the l->tail->next after pop points to NULL

disconnect3d commented on double_linked_list/double_linked_list.c:
```diff
@@ -3,22 +3,34 @@
+        if (l->quantity == 1)
+            free(first);
+        else if (l->quantity == 0)
+            return;
```
Imho this is too complex. Also both `if (l->head == NULL)` and `else` paths does the same thing - `free(l)`. Sometimes such redundancy is okay, but here it is totally redundant. I wonder if a smart compiler would change this to:

``` C
if (l->head != NULL) { ... }
free(l);
```

...but probably not.

In the end I would rewrite the `list_free` function to:

``` C
void list_free(list* l) {
    if (l->quantity > 0) {
        node* current = l->head;
        node* next = l->head->next;

        while(next) {
            free(current);
            current = next;
            next = next->next;
        }

        // deallocating last element
        free(current);
    }

    free(l);
}
```

disconnect3d commented on double_linked_list/double_linked_list.c:
```diff
@@ -0,0 +1,153 @@
+    if (!l->head->next)
+        l->tail = l->head;
+    l->quantity += 1;
+    return true;
```
Consider list, where `list->head` and `list->tail` are equal, and points to `node0`:

```
node0 [HEAD, TAIL]:
+------+------+------+
| PREV | DATA | NEXT |
+------+------+------+
| NULL |   0  | NULL |
+------+------+------+
```

Going through the code, we have:
1. New node allocation, initialization.
2. Interesting code:
```
    if (l->head)
        l->head->prev = new_node;
    l->head = new_node;
    if (!l->head->next)
        l->tail = l->head;
```
After firing it on above list example, both the `list->tail` and `list->head` points to the first node (which is `new_node`).

disconnect3d commented on double_linked_list/double_linked_list.c:
```diff
@@ -0,0 +1,153 @@
+        l->tail = tmp;
+        node* iter = l->head;
+        node* temp = l->head->prev;
+        void* holder;
```
Why `void*` when it stores `node*` all the time?

You rarely need to use void pointers.

disconnect3d commented on double_linked_list/double_linked_list.c:
```diff
@@ -0,0 +1,153 @@
+        chosen_el = iter->data;
+    }
+    else if (pos > quantity/2){
+        int counter = 0;
```
Add a testcase when this branch is taken.

For a list of `{0, 1, 2, 3, 4, 5, 6}`, if I call `list_get_el(&list, 5)` instead of `5` I will get `0`.

Commit:
Changed list_free so it doesn't have redundant code
Changed void pointer to node* in list_reverse
Fixed list_get_el and added a testcase for when the given position is
bigger than number of elements divided by two

Commit:
Added some newlines
